---
title: "Project_14624"
author: "RK"
date: "06/21/2023"
output: html_document
---
shell scripts ran on mski2630
R ran on lski2753

CUT&RUN For Young, SG, CD

Moved YJ data to separate project folder, only the WT samples, IDR.
  Transfer relevant files from: bam_dovetailON , MACS2_BAMPE_narrowPeak_cutoffAnalysis_SPMR_dovetailON_keepdupall, MACS2_Pooled_dovetailON_keepdupall, tracks_bamCov


Run moveFastqFilesFromServer_v3.sh
Run runFastqcCNRTrimAndrunFastqcAndAlignAndMarkDup_dovetailONandOFF_andSpikeIn_CNR_RKv1.sh
Generate sampletable / IP pair table
Run indivPeakCall_CUTANDRUN_dovetailON.sh


```{r}
library(stringr)
```
Dovetail ON
## sampletable
```{r}
bam.files <- list.files(path = "bam_dovetailON", pattern = ".bam$", full.names = TRUE)
# bam.sg <- bam.files[c(1:7)]
bam.y <- bam.files[c(8:12, 16:25)]
# bam.cd <- bam.files[c(13:15, 26:31)]
```

Y
```{r}
st <- stringr::str_split(bam.y, "_", simplify = TRUE)[,c(3:8)]
st <- as.data.frame(st)
colnames(st) <- c("genotype", "cellcount", "antibody", "condition", "abtype", "replicate")
st$pairID <- paste0(st$genotype, "_", st$cellcount, "_", st$antibody, "_", st$condition, "_", st$abtype)
st$fileName <- bam.y
row.names(st) <- paste0(st$genotype, "_", st$cellcount, "_", st$antibody, "_", st$condition, "_", st$abtype, "_", st$replicate)
saveRDS(st, "sampletable_dovetailON_Y.rds")
```


```{r}
# st.sg <- readRDS("sampletable_dovetailON_SG.rds")
st.y <- readRDS("sampletable_dovetailON_Y.rds")
# st.cd <- readRDS("sampletable_dovetailON_CD.rds")
```

```{r}
# tabdf.sg <- data.frame(INPUT = c(rep(st.sg$fileName[which(st.sg$antibody %in% "input")], 6)),
#                     IP = st.sg$fileName[-which(st.sg$antibody %in% "input")])
tabdf.y <- data.frame(INPUT = c(rep(st.y$fileName[which(st.y$antibody %in% "input")], 14)),
                    IP = st.y$fileName[-which(st.y$antibody %in% "input")])
# tabdf.cd <- data.frame(INPUT = c(rep(st.cd$fileName[which(st.cd$antibody %in% "input" & st.cd$maturation %in% "proB")], 3), 
#                                  rep(st.cd$fileName[which(st.cd$antibody %in% "input" & st.cd$maturation %in% "matB")], 4)),
#                     IP = c(st.cd$fileName[-which(st.cd$antibody %in% "input")]))
```


```{r}
# st.ips <- st[which(st$antibody %in% "SigmaPolII"),]
# st.inputs <-  st[which(st$antibody %in% "input"),]
```

## IP pairs
```{r}
# tabdf <- data.frame(INPUT = c(rep(st.inputs$fileName[1], 4), rep(st.inputs$fileName[2], 4)),
#                     IP = st.ips$fileName)
tabdf <- rbind(tabdf.sg, tabdf.y, tabdf.cd)
write.table(tabdf, "input_IP_pairs_dovetailON.txt", quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
```

***
#**Size factors - Spike-in**
***
Use SacCer spike in DNA to determine sizefactors, using C as 10000 (https://yezhengstat.github.io/CUTTag_tutorial/#V_Spike-in_calibration)

```{r}
st <- readRDS("sampletable_dovetailON_Y.rds")

st$sampleName <- stringr::str_split(st$fileName, "\\/", simplify = TRUE)[,2]
st$sampleName <- stringr::str_split(st$sampleName, "_dupMark", simplify = TRUE)[,1]

# st$bwFileName <- stringr::str_split(st$fileName, "\\/", simplify = TRUE)[,2]
# st$bwFileName <- stringr::str_split(st$bwFileName, "\\.", simplify = TRUE)[,1]
# st$bwFileName <- paste0("tracks_bamCov/bam_dovetailON/", st$bwFileName, ".bw")
# file.exists(st$bwFileName)

spikein <- read.table("spikein_bam/info.log")[,c(2,8)]
spikein$V8 <- as.numeric(stringr::str_split(spikein$V8, ",", simplify = TRUE)[,1])
spikein$V2 <- stringr::str_split(spikein$V2, "\\/", simplify = TRUE)[,2]
colnames(spikein) <- c("sampleName", "spikeinReads")
st <- merge(st,spikein, by = "sampleName", all.x = TRUE)
# #only keep SG's
# st <- st[which(st$antibody %in% "STAT4"), ]
# spikein <- spikein[grep("STAT4", spikein$V2), ]

# !!DOUBLE CHECK if order is same between st and spikein!!
# sapply(spikein$V2, function(x){
#   grep(x, st$fileName)
# })
#
# st$spikeinReads <- spikein$V8
const <- 10000
st$sizeFactors <- const/st$spikeinReads

sfdf <- data.frame(row.names = rownames(st),
                   "fileName" = st$fileName,
                   "sf" = st$sizeFactors)
saveRDS(sfdf, "sizeFactors_inverse_SacCer_Y.rds")
write.table(sfdf,file="sizeFactors_inverse_SacCer_Y.txt",row.names=T,col.names=F,quote=F,sep="\t")
```



